File: 2labz/2labz1/1.js
function minOperations(target) {
    const n = target.length;
    const dp = new Array(n + 1).fill(0);
    let usedDouble = false;
    
    // Заполняем массив динамического программирования
    for (let i = 1; i <= n; i++) {
        dp[i] = dp[i-1] + 1; // Базовый случай
        
        // Проверяем возможность единственного удвоения
        if (!usedDouble && i % 2 === 0 && target.slice(0, i/2) === target.slice(i/2, i)) {
            if (dp[i/2] + 1 < dp[i]) {
                dp[i] = dp[i/2] + 1;
                usedDouble = true;
            }
        }
    }

    // Восстанавливаем последовательность
    const sequence = [];
    let i = n;
    let doubleUsed = false;
    
    while (i > 0) {
        // Определяем тип последней операции
        if (!doubleUsed && i % 2 === 0 && 
            target.slice(0, i/2) === target.slice(i/2, i) && 
            dp[i] === dp[i/2] + 1) {
            sequence.push(`Удвоение (получено '${target.slice(0, i)}')`);
            i /= 2;
            doubleUsed = true;
        } else {
            sequence.push(`Добавлен '${target[i-1]}'`);
            i--;
        }
    }
    
    return {
        steps: dp[n],
        sequence: sequence.reverse() // Разворачиваем последовательность
    };
}

const readline = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout
});

readline.question('Введите целевую строку: ', (target) => {
    const result = minOperations(target);
    console.log(`Минимальное количество операций: ${result.steps}`);
    console.log('Последовательность:');
    result.sequence.forEach((step, i) => {
        console.log(`${i + 1}. ${step}`);
    });
    readline.close();
});

File: 2labz/2labz1/MyCSharpApp/1.cs
using System;
using System.Collections.Generic;

// Основная программа
class Program
{
    static void Main()
    {
        // Ввод целевой строки пользователем
        Console.Write("Введите целевую строку: ");
        string target = Console.ReadLine() ?? string.Empty; // Чтение строки или пустая строка по умолчанию
        
        // Вызов функции для получения минимальных операций и последовательности
        var result = MinOperations(target);
        
        // Вывод минимального количества операций
        Console.WriteLine($"Минимальное количество операций: {result.steps}");
        Console.WriteLine("Последовательность:");
        // Вывод каждого шага последовательности
        for (int i = 0; i < result.sequence.Count; i++)
        {
            Console.WriteLine($"{i + 1}. {result.sequence[i]}");
        }
    }

    static (int steps, List<string> sequence) MinOperations(string target)
    {
        int n = target.Length; // Длина целевой строки
        // Используем List вместо массива
        List<int> dp = new List<int>(n + 1); // Создание списка для хранения минимальных операций
        for (int i = 0; i <= n; i++) dp.Add(i); // Инициализация значениями 0..n
        bool usedDouble = false; // Флаг, указывающий использование удвоения

        // Основной цикл для заполнения dp
        for (int i = 1; i <= n; i++)
        {
            dp[i] = dp[i - 1] + 1; // Минимум операций — добавить один символ
            
            // Проверка возможности удвоения
            if (!usedDouble && i % 2 == 0)
            {
                bool canDouble = true; // Флаг, можно ли удвоить
                int half = i / 2; // Индекс половины строки
                // Проверка совпадения половин строки
                for (int j = 0; j < half; j++)
                {
                    if (target[j] != target[half + j])
                    {
                        canDouble = false; // Если символы не совпадают, удвоение невозможно
                        break;
                    }
                }
                // Если можно удвоить и это уменьшает число операций
                if (canDouble && dp[half] + 1 < dp[i])
                {
                    dp[i] = dp[half] + 1; // Обновляем минимальное число операций
                    usedDouble = true; // Устанавливаем флаг, что удвоение было использовано
                }
            }
        }

        // Восстановление последовательности шагов
        List<string> sequence = new List<string>(); // Создаем список для хранения шагов
        int pos = n; // Текущий индекс в строке
        bool doubleUsed = false; // Флаг использования удвоения при восстановлении

        while (pos > 0)
        {
            // Проверка возможности удвоения при обратном восстановлении
            if (!doubleUsed && pos % 2 == 0)
            {
                int half = pos / 2; // Индекс половины
                bool canDouble = true; // Можно ли удвоить
                for (int j = 0; j < half; j++)
                {
                    if (target[j] != target[half + j])
                    {
                        canDouble = false;
                        break;
                    }
                }
                // Если удвоение соответствует минимальному числу операций
                if (canDouble && dp[pos] == dp[half] + 1)
                {
                    // Добавляем шаг удвоения
                    sequence.Insert(0, $"Удвоение (получено '{target.Substring(0, pos)}')");
                    pos /= 2; // Переход к предыдущему состоянию
                    doubleUsed = true; // Помечаем, что удвоение использовано
                    continue; // Продолжаем цикл
                }
            }
            // Добавление шага добавления символа
            sequence.Insert(0, $"Добавлен '{target[pos - 1]}'");
            pos--; // Переход к предыдущему символу
        }

        // Возвращение минимальных операций и последовательности
        return (dp[n], sequence);
    }
}
